File: backend/app/models/autopilot.py
"""
DealMotion Autopilot - Pydantic Models
SPEC-045 / TASK-048

Models for the Autopilot feature: proposals, settings, outcomes, and preferences.
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any, Literal
from datetime import datetime
from enum import Enum


# =============================================================================
# ENUMS
# =============================================================================

class ProposalType(str, Enum):
    """Types of autopilot proposals."""
    RESEARCH_PREP = "research_prep"       # New meeting, unknown org
    PREP_ONLY = "prep_only"               # Known prospect, no prep
    FOLLOWUP_PACK = "followup_pack"       # Post-meeting
    REACTIVATION = "reactivation"         # Silent prospect
    COMPLETE_FLOW = "complete_flow"       # Research done, no next step


class ProposalStatus(str, Enum):
    """Status of an autopilot proposal."""
    PROPOSED = "proposed"
    ACCEPTED = "accepted"
    EXECUTING = "executing"
    COMPLETED = "completed"
    DECLINED = "declined"
    SNOOZED = "snoozed"
    EXPIRED = "expired"
    FAILED = "failed"


class TriggerType(str, Enum):
    """What triggered the proposal."""
    CALENDAR_NEW_ORG = "calendar_new_org"
    CALENDAR_KNOWN_PROSPECT = "calendar_known_prospect"
    MEETING_ENDED = "meeting_ended"
    TRANSCRIPT_READY = "transcript_ready"
    PROSPECT_SILENT = "prospect_silent"
    FLOW_INCOMPLETE = "flow_incomplete"
    MANUAL = "manual"


class NotificationStyle(str, Enum):
    """Notification frequency preference."""
    EAGER = "eager"         # Notify immediately
    BALANCED = "balanced"   # Smart timing
    MINIMAL = "minimal"     # Only urgent


class OutcomeRating(str, Enum):
    """Meeting outcome rating."""
    POSITIVE = "positive"
    NEUTRAL = "neutral"
    NEGATIVE = "negative"


class OutcomeSource(str, Enum):
    """Source of outcome rating."""
    USER_INPUT = "user_input"
    FOLLOWUP_SENTIMENT = "followup_sentiment"
    INFERRED = "inferred"


class PrepLength(str, Enum):
    """Preferred prep length."""
    SHORT = "short"
    MEDIUM = "medium"
    LONG = "long"


class LunaMode(str, Enum):
    """Luna greeting modes."""
    MORNING = "morning"
    PROPOSAL = "proposal"
    URGENCY = "urgency"
    CELEBRATION = "celebration"
    COACH = "coach"


# =============================================================================
# PROPOSAL MODELS
# =============================================================================

class SuggestedAction(BaseModel):
    """A single suggested action within a proposal."""
    action: str  # 'research', 'prep', 'followup_summarize', 'followup_email'
    params: Dict[str, Any] = Field(default_factory=dict)


class AutopilotProposalBase(BaseModel):
    """Base model for autopilot proposals."""
    proposal_type: ProposalType
    trigger_type: TriggerType
    trigger_entity_id: Optional[str] = None
    trigger_entity_type: Optional[str] = None
    title: str
    description: Optional[str] = None
    luna_message: str
    suggested_actions: List[SuggestedAction] = Field(default_factory=list)
    priority: int = Field(default=50, ge=0, le=100)
    expires_at: Optional[datetime] = None
    context_data: Dict[str, Any] = Field(default_factory=dict)


class AutopilotProposalCreate(AutopilotProposalBase):
    """Model for creating a new proposal."""
    organization_id: str
    user_id: str


class AutopilotProposal(AutopilotProposalBase):
    """Full autopilot proposal model."""
    id: str
    organization_id: str
    user_id: str
    status: ProposalStatus = ProposalStatus.PROPOSED
    decided_at: Optional[datetime] = None
    decision_reason: Optional[str] = None
    snoozed_until: Optional[datetime] = None
    execution_started_at: Optional[datetime] = None
    execution_completed_at: Optional[datetime] = None
    execution_result: Optional[Dict[str, Any]] = None
    execution_error: Optional[str] = None
    artifacts: List[Dict[str, Any]] = Field(default_factory=list)
    expired_reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ProposalActionRequest(BaseModel):
    """Request model for proposal actions (accept/decline/snooze)."""
    reason: Optional[str] = None
    snooze_until: Optional[datetime] = None


class ProposalCounts(BaseModel):
    """Counts of proposals by status."""
    proposed: int = 0
    executing: int = 0
    completed: int = 0
    declined: int = 0
    snoozed: int = 0
    expired: int = 0
    failed: int = 0


class ProposalsResponse(BaseModel):
    """Response model for listing proposals."""
    proposals: List[AutopilotProposal]
    counts: ProposalCounts
    total: int


# =============================================================================
# SETTINGS MODELS
# =============================================================================

class AutopilotSettingsBase(BaseModel):
    """Base model for autopilot settings."""
    enabled: bool = True
    auto_research_new_meetings: bool = True
    auto_prep_known_prospects: bool = True
    auto_followup_after_meeting: bool = True
    reactivation_days_threshold: int = 14
    prep_hours_before_meeting: int = 24
    notification_style: NotificationStyle = NotificationStyle.BALANCED
    excluded_meeting_keywords: List[str] = Field(default_factory=list)


class AutopilotSettingsUpdate(BaseModel):
    """Model for updating autopilot settings."""
    enabled: Optional[bool] = None
    auto_research_new_meetings: Optional[bool] = None
    auto_prep_known_prospects: Optional[bool] = None
    auto_followup_after_meeting: Optional[bool] = None
    reactivation_days_threshold: Optional[int] = None
    prep_hours_before_meeting: Optional[int] = None
    notification_style: Optional[NotificationStyle] = None
    excluded_meeting_keywords: Optional[List[str]] = None


class AutopilotSettings(AutopilotSettingsBase):
    """Full autopilot settings model."""
    id: str
    user_id: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# =============================================================================
# LUNA GREETING MODELS
# =============================================================================

class LunaGreeting(BaseModel):
    """Model for Luna's dynamic greeting."""
    mode: LunaMode
    message: str
    emphasis: Optional[str] = None
    action: Optional[str] = None
    action_route: Optional[str] = None
    pending_count: int = 0
    urgent_count: int = 0


class UpcomingMeeting(BaseModel):
    """Model for upcoming meeting info."""
    id: str
    title: str
    company: Optional[str] = None
    start_time: datetime
    starts_in_hours: float
    has_prep: bool = False
    prospect_id: Optional[str] = None


# =============================================================================
# STATS MODELS
# =============================================================================

class AutopilotStats(BaseModel):
    """Stats for the autopilot home page."""
    pending_count: int = 0
    urgent_count: int = 0
    completed_today: int = 0
    upcoming_meetings: List[UpcomingMeeting] = Field(default_factory=list)
    luna_greeting: LunaGreeting


# =============================================================================
# OUTCOME MODELS
# =============================================================================

class MeetingOutcomeBase(BaseModel):
    """Base model for meeting outcomes."""
    calendar_meeting_id: Optional[str] = None
    preparation_id: Optional[str] = None
    followup_id: Optional[str] = None
    prospect_id: Optional[str] = None
    outcome_rating: Optional[OutcomeRating] = None
    outcome_source: Optional[OutcomeSource] = None
    prep_viewed: bool = False
    prep_view_duration_seconds: Optional[int] = None
    prep_scroll_depth: Optional[float] = None
    had_contact_analysis: Optional[bool] = None
    had_kb_content: Optional[bool] = None
    prep_length_words: Optional[int] = None


class MeetingOutcomeCreate(MeetingOutcomeBase):
    """Model for creating a meeting outcome."""
    organization_id: str
    user_id: str


class MeetingOutcome(MeetingOutcomeBase):
    """Full meeting outcome model."""
    id: str
    organization_id: str
    user_id: str
    created_at: datetime

    class Config:
        from_attributes = True


class OutcomeRequest(BaseModel):
    """Request model for recording an outcome."""
    outcome_rating: OutcomeRating
    calendar_meeting_id: Optional[str] = None
    preparation_id: Optional[str] = None
    followup_id: Optional[str] = None
    prospect_id: Optional[str] = None


# =============================================================================
# PREFERENCES MODELS
# =============================================================================

class UserPrepPreferencesBase(BaseModel):
    """Base model for user prep preferences."""
    preferred_length: Optional[PrepLength] = None
    valued_sections: List[str] = Field(default_factory=list)
    deemphasized_sections: List[str] = Field(default_factory=list)
    avg_prep_view_duration_seconds: Optional[int] = None
    prep_completion_rate: Optional[float] = None
    positive_outcome_rate: Optional[float] = None
    sample_size: int = 0


class UserPrepPreferences(UserPrepPreferencesBase):
    """Full user prep preferences model."""
    id: str
    user_id: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# =============================================================================
# PREP VIEWED TRACKING
# =============================================================================

class PrepViewedRequest(BaseModel):
    """Request model for tracking prep viewed."""
    preparation_id: str
    view_duration_seconds: int
    scroll_depth: float = Field(ge=0.0, le=1.0)


# =============================================================================
# LUNA MESSAGE TEMPLATES
# =============================================================================

LUNA_TEMPLATES = {
    # New meeting, unknown org
    "research_prep_new": "Je hebt {time} een meeting met een nieuw bedrijf. Wil je dat ik research doe en een briefing maak?",
    
    # Known prospect, no prep
    "prep_only": "Je hebt al research over {company}. Wil je dat ik een meeting prep maak voor {time}?",
    
    # Post-meeting
    "followup_pack": "Je meeting met {company} is afgelopen. Ik kan een samenvatting en follow-up email maken.",
    
    # Silent prospect
    "reactivation": "{company} is al {days} dagen stil. Het laatste gesprek was positief. Tijd voor een check-in?",
    
    # Incomplete flow
    "complete_flow": "Je hebt {contact} toegevoegd aan {company}. Wil je een prep maken voor het eerste gesprek?",
    
    # Execution complete
    "execution_complete": "Je briefing voor {company} is klaar!",
    
    # Execution failed
    "execution_failed": "Er ging iets mis bij het maken van je briefing voor {company}. Wil je het opnieuw proberen?",
    
    # Urgent
    "urgent": "Let op: je meeting met {company} begint over {hours} uur!",
}

File: frontend/types/autopilot.ts
/**
 * DealMotion Autopilot - TypeScript Types
 * SPEC-045 / TASK-048
 */

// =============================================================================
// ENUMS
// =============================================================================

export type ProposalType =
  | 'research_prep'      // New meeting, unknown org
  | 'prep_only'          // Known prospect, no prep
  | 'followup_pack'      // Post-meeting
  | 'reactivation'       // Silent prospect
  | 'complete_flow'      // Research done, no next step

export type ProposalStatus =
  | 'proposed'
  | 'accepted'
  | 'executing'
  | 'completed'
  | 'declined'
  | 'snoozed'
  | 'expired'
  | 'failed'

export type TriggerType =
  | 'calendar_new_org'
  | 'calendar_known_prospect'
  | 'meeting_ended'
  | 'transcript_ready'
  | 'prospect_silent'
  | 'flow_incomplete'
  | 'manual'

export type NotificationStyle = 'eager' | 'balanced' | 'minimal'

export type OutcomeRating = 'positive' | 'neutral' | 'negative'

export type LunaMode = 'morning' | 'proposal' | 'urgency' | 'celebration' | 'coach'


// =============================================================================
// PROPOSAL
// =============================================================================

export interface SuggestedAction {
  action: string
  params: Record<string, unknown>
}

export interface ProposalArtifact {
  type: 'research' | 'prep' | 'followup' | 'redirect'
  id?: string
  route?: string
  message?: string
}

export interface AutopilotProposal {
  id: string
  organization_id: string
  user_id: string
  proposal_type: ProposalType
  trigger_type: TriggerType
  trigger_entity_id: string | null
  trigger_entity_type: string | null
  title: string
  description: string | null
  luna_message: string
  proposal_reason: string | null  // Why this proposal was created
  suggested_actions: SuggestedAction[]
  status: ProposalStatus
  priority: number
  decided_at: string | null
  decision_reason: string | null
  snoozed_until: string | null
  execution_started_at: string | null
  execution_completed_at: string | null
  execution_result: Record<string, unknown> | null
  execution_error: string | null
  artifacts: ProposalArtifact[]
  expires_at: string | null
  expired_reason: string | null
  context_data: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface ProposalCounts {
  proposed: number
  executing: number
  completed: number
  declined: number
  snoozed: number
  expired: number
  failed: number
}

export interface ProposalsResponse {
  proposals: AutopilotProposal[]
  counts: ProposalCounts
  total: number
}

export interface ProposalActionRequest {
  reason?: string
  snooze_until?: string
}


// =============================================================================
// SETTINGS
// =============================================================================

export interface AutopilotSettings {
  id: string
  user_id: string
  enabled: boolean
  auto_research_new_meetings: boolean
  auto_prep_known_prospects: boolean
  auto_followup_after_meeting: boolean
  reactivation_days_threshold: number
  prep_hours_before_meeting: number
  notification_style: NotificationStyle
  excluded_meeting_keywords: string[]
  created_at: string
  updated_at: string
}

export interface AutopilotSettingsUpdate {
  enabled?: boolean
  auto_research_new_meetings?: boolean
  auto_prep_known_prospects?: boolean
  auto_followup_after_meeting?: boolean
  reactivation_days_threshold?: number
  prep_hours_before_meeting?: number
  notification_style?: NotificationStyle
  excluded_meeting_keywords?: string[]
}


// =============================================================================
// LUNA GREETING
// =============================================================================

export interface LunaGreeting {
  mode: LunaMode
  message: string
  emphasis: string | null
  action: string | null
  action_route: string | null
  pending_count: number
  urgent_count: number
}

export interface UpcomingMeeting {
  id: string
  title: string
  company: string | null
  start_time: string
  starts_in_hours: number
  has_prep: boolean
  prospect_id: string | null
}


// =============================================================================
// STATS
// =============================================================================

export interface AutopilotStats {
  pending_count: number
  urgent_count: number
  completed_today: number
  upcoming_meetings: UpcomingMeeting[]
  luna_greeting: LunaGreeting
}


// =============================================================================
// OUTCOMES
// =============================================================================

export interface AutopilotMeetingOutcome {
  id: string
  organization_id: string
  user_id: string
  calendar_meeting_id: string | null
  preparation_id: string | null
  followup_id: string | null
  prospect_id: string | null
  outcome_rating: OutcomeRating | null
  outcome_source: 'user_input' | 'followup_sentiment' | 'inferred' | null
  prep_viewed: boolean
  prep_view_duration_seconds: number | null
  prep_scroll_depth: number | null
  had_contact_analysis: boolean | null
  had_kb_content: boolean | null
  prep_length_words: number | null
  created_at: string
}

export interface OutcomeRequest {
  outcome_rating: OutcomeRating
  calendar_meeting_id?: string
  preparation_id?: string
  followup_id?: string
  prospect_id?: string
}

export interface PrepViewedRequest {
  preparation_id: string
  view_duration_seconds: number
  scroll_depth: number
}


// =============================================================================
// CONTEXT
// =============================================================================

export interface AutopilotContextValue {
  // State
  isLoading: boolean
  isEnabled: boolean
  proposals: AutopilotProposal[]
  counts: ProposalCounts
  stats: AutopilotStats | null
  settings: AutopilotSettings | null
  
  // Actions
  refreshProposals: () => Promise<void>
  acceptProposal: (id: string, reason?: string) => Promise<void>
  completeProposal: (id: string) => Promise<void>  // For inline actions (skip Inngest)
  declineProposal: (id: string, reason?: string) => Promise<void>
  snoozeProposal: (id: string, until: Date, reason?: string) => Promise<void>
  retryProposal: (id: string) => Promise<void>
  refreshStats: () => Promise<void>
  updateSettings: (updates: AutopilotSettingsUpdate) => Promise<void>
  recordOutcome: (outcome: OutcomeRequest) => Promise<void>
  recordPrepViewed: (data: PrepViewedRequest) => Promise<void>
}


// =============================================================================
// UI HELPERS
// =============================================================================

export const PROPOSAL_STATUS_COLORS: Record<ProposalStatus, string> = {
  proposed: 'bg-yellow-100 text-yellow-800',
  accepted: 'bg-blue-100 text-blue-800',
  executing: 'bg-blue-100 text-blue-800',
  completed: 'bg-green-100 text-green-800',
  declined: 'bg-gray-100 text-gray-800',
  snoozed: 'bg-purple-100 text-purple-800',
  expired: 'bg-gray-100 text-gray-500',
  failed: 'bg-red-100 text-red-800',
}

export const PROPOSAL_STATUS_LABELS: Record<ProposalStatus, string> = {
  proposed: 'Actie nodig',
  accepted: 'Geaccepteerd',
  executing: 'Bezig...',
  completed: 'Klaar',
  declined: 'Afgewezen',
  snoozed: 'Uitgesteld',
  expired: 'Verlopen',
  failed: 'Mislukt',
}

export const PROPOSAL_TYPE_ICONS: Record<ProposalType, string> = {
  research_prep: '🔍',
  prep_only: '📋',
  followup_pack: '✉️',
  reactivation: '🔄',
  complete_flow: '➡️',
}

export const LUNA_MODE_ICONS: Record<LunaMode, string> = {
  morning: '☀️',
  proposal: '📬',
  urgency: '⏰',
  celebration: '🎉',
  coach: '💡',
}


// =============================================================================
// SNOOZE OPTIONS
// =============================================================================

export interface SnoozeOption {
  label: string
  getValue: () => Date
}

export const SNOOZE_OPTIONS: SnoozeOption[] = [
  {
    label: '1 hour',
    getValue: () => new Date(Date.now() + 60 * 60 * 1000),
  },
  {
    label: 'This evening',
    getValue: () => {
      const date = new Date()
      date.setHours(20, 0, 0, 0)
      if (date <= new Date()) {
        date.setDate(date.getDate() + 1)
      }
      return date
    },
  },
  {
    label: 'Tomorrow',
    getValue: () => {
      const date = new Date()
      date.setDate(date.getDate() + 1)
      date.setHours(9, 0, 0, 0)
      return date
    },
  },
  {
    label: 'After meeting',
    getValue: () => {
      // Default to 3 hours from now (typical meeting time)
      return new Date(Date.now() + 3 * 60 * 60 * 1000)
    },
  },
  {
    label: 'Next week',
    getValue: () => {
      const date = new Date()
      // Find next Monday
      const daysUntilMonday = (8 - date.getDay()) % 7 || 7
      date.setDate(date.getDate() + daysUntilMonday)
      date.setHours(9, 0, 0, 0)
      return date
    },
  },
  {
    label: 'In 3 days',
    getValue: () => {
      const date = new Date()
      date.setDate(date.getDate() + 3)
      date.setHours(9, 0, 0, 0)
      return date
    },
  },
]
